test/new-e2e/tests/agent-shared-components/forwarder/nss_failover_test.go

```diff
@@ -0,0 +1,336 @@
+	fakeintake1IP, err := hostIPFromURL(v.Env().Fakeintake1.URL())
+	v.NoError(err)
+	setHostEntry(v.T(), v.Env().VM, intakeName, fakeintake1IP)
+
+	// configure agent to use the custom intake, set connection_reset_interval, use logs, and processes
+	agentOptions := []agentparams.Option{
+		agentparams.WithAgentConfig(agentConfig),
+		agentparams.WithLogs(),
+		agentparams.WithIntakeHostname(intakeName),
+		agentparams.WithIntegration("custom_logs.d", customLogsConfig),
+	}
+	v.UpdateEnv(multiFakeintakeStackDef(agentOptions...))
+
+	// check that fakeintake1 is used as intake and not fakeintake2
+	v.T().Logf("checking that the agent contacts main intake at %s", fakeintake1IP)
+	v.requireIntakeIsUsed(v.Env().Fakeintake1, intakeMaxWaitTime, intakeTick)
+	v.requireIntakeNotUsed(v.Env().Fakeintake2, intakeMaxWaitTime, intakeTick)
+
+	// perform NSS change
+	fakeintake2IP, err := hostIPFromURL(v.Env().Fakeintake2.URL())
+	v.NoError(err)
+	setHostEntry(v.T(), v.Env().VM, intakeName, fakeintake2IP)
+
+	// check that fakeintake2 is used as intake and not fakeintake1
+	v.T().Logf("checking that the agent contacts fallback intake at %s", fakeintake2IP)
+	intakeMaxWaitTime := connectionResetInterval*time.Second + intakeMaxWaitTime
+	v.requireIntakeIsUsed(v.Env().Fakeintake2, intakeMaxWaitTime, intakeTick)
+	v.requireIntakeNotUsed(v.Env().Fakeintake1, intakeMaxWaitTime, intakeTick)
+}
+
+// requireIntakeIsUsed checks that the given intakes receives metrics, logs, and flares
+func (v *multiFakeIntakeSuite) requireIntakeIsUsed(intake *client.Fakeintake, intakeMaxWaitTime, intakeTick time.Duration) {
+	checkFn := func(t *assert.CollectT) {
+		// check metrics
+		metricNames, err := intake.GetMetricNames()
+		require.NoError(t, err)
+		assert.NotEmpty(t, metricNames)
+
+		// check logs
+		v.Env().VM.Execute(fmt.Sprintf("echo 'totoro' >> %s", logFile))
+		logs, err := intake.FilterLogs(logService)
+		require.NoError(t, err)
+		assert.NotEmpty(t, logs)
+
+		// check flares
+		v.Env().Agent.Flare(client.WithArgs([]string{"--email", "e2e@test.com", "--send"}))
+		_, err = intake.GetLatestFlare()
+		if err != nil {
+			require.ErrorIs(t, err, fi.ErrNoFlareAvailable)
+		}
+		assert.NoError(t, err)
+	}
+
+	require.EventuallyWithT(v.T(), checkFn, intakeMaxWaitTime, intakeTick)
+}
+
+// requireIntakeNotUsed checks that the given intake doesn't receive metrics, logs, and flares
+func (v *multiFakeIntakeSuite) requireIntakeNotUsed(intake *client.Fakeintake, intakeMaxWaitTime, intakeTick time.Duration) {
+	checkFn := func(t *assert.CollectT) {
+		// flush intake
+		intake.FlushServerAndResetAggregators()
+
+		// write a log
+		v.Env().VM.Execute(fmt.Sprintf("echo 'totoro' >> %s", logFile))
+
+		// send a flare
+		v.Env().Agent.Flare(client.WithArgs([]string{"--email", "e2e@test.com", "--send"}))
+
+		// give time to the agent to send things
+		time.Sleep(intakeUnusedWaitTime)
+
+		stats, err := intake.RouteStats()
+		require.NoError(t, err)
+
+		assert.Empty(t, stats)
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
This could move out of the eventually. The first time this fails, it is an actual bug, not a flake.

Reply:
Not quite, because the agent takes some time (at most 2 minutes in theory) to notice that the intake changed and redirect connections, so the first intake can still receive payloads for the first few minutes.

I could also rewrite that part to remove the `EventuallyWithT` by waiting 2 minutes (+ possibly some delta to minimize risks) and then requiring.